pages/blog/why-not-discordjs.mdx

```diff
@@ -0,0 +1,170 @@
+
+<Adsense />
+
+![discord.js make cache](@static/blog/why-not-discordjs/make-cache.png)
+
+#### discord.js-light
+
+幸好有人發現了這個問題並製作出 [discord.js-light](https://www.npmjs.com/package/discord.js-light) 套件，可以直接覆蓋掉幾乎所有的 Manager 設定，不過因為還是第三方軟體，並不是 discord.js 自帶功能，最好在使用之前注意會不會讓程式崩壞。
+
+![discord.js-light npm page](@static/blog/why-not-discordjs/discordjs-light.png)
+
+我自己有發現了一個神奇但是挺好用的解法，可以直接使用 JavaScript 神奇的 [ProtoType](https://developer.mozilla.org/zh-TW/docs/Web/JavaScript/Inheritance_and_the_prototype_chain) 特性，
+
+舉個例子：若你發現你真的不會使用到 `ChannelManager#cache` 裡的東西，先將 `bot.channels` 的 `set` 函式清空，需要用到時使用 `ChannelManager#cache` 的原型 `Map` 來執行 `set` 的工作，
+
+```js
+const bot = new Client({ ... })
+
+// make the set function does nothing
+bot.channels.cache.set = () => {}
+
+bot.on("ready", () => {
+	console.log(bot.channels.cache.size) // 0
+})
+
+// resolver example
+async function resolveChannel(channelId) {
+	if(bot.channels.has(channelId))
+		return bot.channels.get(channelId)
+	
+	const resolved = await bot.channels.fetch(channelId)
+
+	Map.prototype.set.call(bot.channels, channelId, resolved)
+
+	return resolved
+}
+```
+
+可以看到機器人在準備完成後，預先快取的頻道數量是 `0`，而當我們需要手動抓取頻道的時候，使用 `resolveChannel` 函式去手動執行快取的操作，雖然這不是個最佳解，不過 JavaScript 是個神奇的語言所以... 恩！
+
+<Adsense />
+
+### 機器人規模成長後難維護
+
+當機器人成長到一定的大小後，如何架構將是一個重要的課題，架構層面我們先從如何做解耦 (decouple) 開始，可以把它簡單的理解成兩個元件要如何做拆解，詳細的架構說明可以參考 [Day 23: 元件原則 — 耦合性 by JC](https://ithelp.ithome.com.tw/articles/10279202)。
+
+Discord 機器人通常可以拆解為 Gateway 和 Discord 溝通或是接收 Discord 的事件，REST API Proxy 執行發送訊息等操作，而到我最後使用 v13 版本時，所有功能還是都包在一起的，可能他們有在嘗試做一些改變不過我不太確定。
+
+![discord.js Github repo](@static/blog/why-not-discordjs/discordjs-repo.png)
+
+緊耦合對維護來說非常困難，機器人在上線時會依照伺服器數量來分配分片 (Sharding) 的數量，每個分片上線都要先等待五秒的冷卻時間，若每次更新需要重開時，都需要經過一次這樣的上線流程，
+
+而造成的下線時間 (Downtime) 十幾分鐘可是對穩定性和可靠性非常大的致命傷，~~可以看我們 Google 搜尋結果都可以被洗成這樣了...~~
+
+![Yeecord Google SERP](@static/blog/why-not-discordjs/yeecord-search-result.png)
+
+## 那我有其他的選擇嗎？
+
+Create-React-App 是個很棒的 React 工具，用簡單的 `react-scripts` 指令就可以跑起 React，而不用擔心後面 babel、webpack 繁雜的設定等等，
+
+<Adsense />
+
+而在時代不斷的更迭到了現在 2023 年，我們有了更多如 vite、NextJS 等的新選擇，在速度與可擴展性上都有大幅的提升，從前端框架的現象我們也可以來觀察一下，JavaScript 是否還有其他解決方案來製作 Discord 機器人。
+
+### 先看一下函式庫比較圖表
+
+[Discord API Library Comparisons](https://libs.advaith.io/#javascript) 這網站統計了大多數用來呼叫 Discord API 的套件，我們將會專注在套件的支援程度上面，
+
+可以看到除了 discord.js 之外，其他套件也不惶多讓，下面將來講一下我們觀察到的結果。
+
+![Discord bot library in JavaScript](@static/blog/why-not-discordjs/discord-bot-library-javascript.png)
+
+### Discordeno
+
+![Discordeno](@static/blog/why-not-discordjs/discordeno.png)
+
+[Discordeno](https://github.com/discordeno/discordeno) 是使用 deno 編寫的 Discord API 套件，並且在 NodeJS 也可以使用，從他們的 GitHub 首頁來看，上面列舉了許多讓人心動的優勢，讓我們能夠信任並且把機器龍轉移過去：

```

Comment:
Same reason as Vite, I think capitalizing will make it look better.

Reply:
fixed